core/species.cpp

```diff
@@ -4999,24 +5000,44 @@ void Species::WritePopulationTable(tsk_table_collection_t *p_tables)
 				
 				if (CheckSLiMPopulationMetadata(tsk_population_object.metadata, tsk_population_object.metadata_length) == -1)
 				{
-					// The metadata present, if any, is not SLiM metadata, so it should be carried over; note that
-					// SLiM metadata for non-extant subpops gets removed at write time for consistency.  See issue #317.
-					tsk_population_id = tsk_population_table_add_row(
-							&p_tables->populations,
+					// The metadata present, if any, is not SLiM metadata, so it should be carried over.
+					new_metadata_str = tsk_population_object.metadata;
+					new_metadata_length = tsk_population_object.metadata_length;
+				}
+				else
+				{
+					// SLiM metadata for non-extant subpops gets removed at write time for consistency.
+					// See issue #317 for discussion.  However, we keep *names* from SLiM populations
+					// that are found in the table and have been removed, because names provide a
+					// useful way for the user to check that everything is as they expect,
+					// lets them find particular populations without error-prone bookkeeping,
+					// and is the basis for the more user-friendly interfaces in msprime
+					// (e.g., it's important to have names when setting up a more complex model
+					// to use in recapitation). At present, the only way that
+					// entries can have SLiM metadata in the population table
+					// but not correspond to extant populations is if the populations were present
+					// in a tree sequence that was loaded in to SLiM but they were
+					// subsequently removed.
+					std::string metadata_string(
 							tsk_population_object.metadata,
 							tsk_population_object.metadata_length);
-					if (tsk_population_id < 0) handle_error("tsk_population_table_add_row", tsk_population_id);
-					write_empty_row = false;
+					nlohmann::json old_metadata;
+					old_metadata = nlohmann::json::parse(metadata_string);
+					nlohmann::json new_metadata = nlohmann::json::object();
+					if (old_metadata.contains("name")) {
+						new_metadata["name"] = old_metadata["name"];
+						std::string metadata_rec = new_metadata.dump();
+						new_metadata_str = (const char *)metadata_rec.c_str();
```

Comment:
Unfortunately for this elegant solution, this is unsafe.  `new_metadata` goes out of scope at the end of the block, and the pointer taken from `c_str()` here becomes a pointer into an object that no longer exists.  If you're lucky the memory stays valid until `new_metadata_str` is used below, but there is no guarantee of that; this is "undefined behavior", which means whether it works or whether it crashes is undefined.  Sorry, needs to be rewritten, probably back to the kind of structure it had before where the call to `tsk_population_table_add_row()` happened within the scope where the new metadata still points to valid data.

Reply:
Hmm, there's a way to rescue the design, I think, by switching from keeping `new_metadata_str` as a C string to keeping it as a std::string (which means the name will get copied, rather than a pointer to the inside of the string object getting taken).  I'm happy to do it.  Give me the green light and I'll accept this PR and fix it on my end.

Reply:
This actually makes sense!